File: ArDemo/ARKController/Metal/ARKMetalController.m
#import "ARKMetalController.h"
#import <Metal/Metal.h>
#import <MetalKit/MetalKit.h>
#import <ARKit/ARKit.h>
#import "Renderer.h"

@interface MTKView () <RenderDestinationProvider>

@end

@interface ARKMetalController () <MTKViewDelegate>


@property (nonatomic, strong) Renderer *renderer;
@property (nonatomic, strong) MTKView *renderView;
@property CGPoint hitTestFocusPoint;

@end

@implementation ARKMetalController

- (instancetype)initWithSesion:(ARSession *)session
{
    self = [super init];
    
    if (self)
    {
        if ([self setupARWithSession:session] == NO)
        {
            return nil;
        }
    }
    
    return self;
}

- (NSArray *)hitTest:(CGPoint)point withType:(ARHitTestResultType)type
{
    return nil;
}

- (matrix_float4x4)cameraProjectionTransform
{
    return matrix_identity_float4x4;
}

- (void)didChangeTrackingState:(ARCamera *)camera
{
    
}

- (id)currentHitTest {
    return nil;
}


- (UIView *)renderView {
    return _renderView;
}

- (void)setShowMode:(ShowMode)mode {
    
}

- (void)setShowOptions:(ShowOptions)options {
    
}

- (BOOL)setupARWithSession:(ARSession *)session
{
    [self setRenderView:[[MTKView alloc] initWithFrame:[[UIScreen mainScreen] bounds] device:MTLCreateSystemDefaultDevice()]];
    [[self renderView] setBackgroundColor:[UIColor clearColor]];
    [[self renderView] setDelegate:self];
    
    if([[self renderView] device] == nil)
    {
        DDLogError(@"Metal is not supported on this device");
        return NO;
    }
    
    [self setRenderer:[[Renderer alloc] initWithSession:session metalDevice:[[self renderView] device] renderDestinationProvider:[self renderView]]];
    
    [[self renderer] drawRectResized:[[self renderView] bounds].size orientation:[[UIApplication sharedApplication] statusBarOrientation]];
    
    return YES;
}

#pragma mark - MTKViewDelegate

- (void)mtkView:(nonnull MTKView *)view drawableSizeWillChange:(CGSize)size
{
    dispatch_async(dispatch_get_main_queue(), ^
    {
        [[self renderer] drawRectResized:size orientation:[[UIApplication sharedApplication] statusBarOrientation]];
    });
}

- (void)drawInMTKView:(nonnull MTKView *)view
{
    [[self renderer] update];
}

@end

File: XRViewer/ARKController/Metal/ARKMetalController.swift
import ARKit
import Metal
import MetalKit

extension MTKView: RenderDestinationProvider {
}

class ARKMetalController: NSObject, ARKControllerProtocol, MTKViewDelegate {
    
    private var session: ARSession
    var renderer: Renderer!
    private var renderView: MTKView?
    private var anchorsNodes: [AnchorNode] = []
    
    private var showMode: ShowMode? {
        didSet {
            updateModes()
        }
    }
    private var showOptions: ShowOptions? {
        didSet {
            updateModes()
        }
    }
    var planes: [UUID : PlaneNode] = [:]
    private var planeHitTestResults: [ARHitTestResult] = []
    private var currentHitTest: HitTestResult?
    private var hitTestFocusPoint = CGPoint.zero
    var previewingSinglePlane: Bool = false
    var focusedPlane: PlaneNode? {
        didSet {
            if focusedPlane == nil {
                oldValue?.geometry?.firstMaterial?.diffuse.contents = UIImage(named: "Models.scnassets/plane_grid1.png")
            }
        }
    }
    var readyToRenderFrame: Bool = true
    var initializingRender: Bool = true
    
    deinit {
        for view in renderView?.subviews ?? [] {
            view.removeFromSuperview()
        }
        renderView?.delegate = nil
        if let _ = renderView {
            renderView = nil
        }
        appDelegate().logger.debug("ARKMetalController dealloc")
    }
    
    required init(sesion session: ARSession, size: CGSize) {
        self.session = session
        super.init()
        if setupAR(with: session) == false {
            print("Error setting up AR Session with Metal")
        }
    }
    
    func update(_ session: ARSession) {
        self.session = session
        if setupAR(with: session) == false {
            print("Error updating AR Session with Metal")
        }
    }
    
    func clean() {
        for (_, plane) in planes {
            plane.removeFromParentNode()
        }
        planes.removeAll()
        
        for anchor in anchorsNodes {
            anchor.removeFromParentNode()
        }
        anchorsNodes.removeAll()
        
        planeHitTestResults = []
    }
    
    func hitTest(_ point: CGPoint, with type: ARHitTestResult.ResultType) -> [ARHitTestResult] {
        if focusedPlane != nil {
            guard let results = session.currentFrame?.hitTest(point, types: type) else { return [] }
            guard let chosenPlane = focusedPlane else { return [] }
            if let anchorIdentifier = planes.someKey(forValue: chosenPlane) {
                let anchor = results.filter { $0.anchor?.identifier == anchorIdentifier }.first
                if let anchor = anchor {
                    return [anchor]
                }
            }
            return []
        } else {
            return session.currentFrame?.hitTest(point, types: type) ?? []
        }
    }

    func updateModes() {
        guard let showMode = showMode else { return }
//        guard let showOptions = showOptions else { return }
        if showMode == ShowMode.urlDebug || showMode == ShowMode.debug {
//            renderView?.showsStatistics = (showOptions.rawValue & ShowOptions.ARStatistics.rawValue) != 0
//            renderView?.debugOptions = (showOptions.rawValue & ShowOptions.ARPoints.rawValue) != 0 ? .showFeaturePoints : []
        } else {
//            renderView?.showsStatistics = false
//            renderView?.debugOptions = []
        }
    }
    
    func didChangeTrackingState(_ camera: ARCamera?) {
    }
    
//    func currentHitTest() -> Any? {
//        return nil
//    }
    
    func setupAR(with session: ARSession) -> Bool {
        renderView = MTKView()
        renderView = MTKView(frame: UIScreen.main.bounds, device: MTLCreateSystemDefaultDevice())
        renderView?.backgroundColor = UIColor.clear
        renderView?.delegate = self
        
        guard let renderView = renderView else {
            appDelegate().logger.error("Error accessing the renderView")
            return false
        }
        guard let device = renderView.device else {
            appDelegate().logger.error("Metal is not supported on this device")
            return false
        }

        renderer = Renderer(session: session, metalDevice: device, renderDestination: renderView)
        renderer.drawRectResized(size: renderView.bounds.size)
        
        return true
    }
    
    
    
    // MARK: - ARKControllerProtocol
    
    func getRenderView() -> UIView! {
        return renderView
    }
    
    func setHitTestFocus(_ point: CGPoint) {
        return
    }
    
    func setShowMode(_ mode: ShowMode) {
        showMode = mode
    }
    
    func setShowOptions(_ options: ShowOptions) {
        showOptions = options
    }
    
    // MARK: - MTKViewDelegate
    
    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
        renderer.drawRectResized(size: size)
    }
    
    func draw(in view: MTKView) {
        guard readyToRenderFrame || initializingRender else {
            return
        }
        renderer.update()
    }
}
